lib/rex/socket.rb

```diff
@@ -809,10 +828,64 @@ def type?
 
 protected
 
-  attr_writer :peerhost, :peerport, :localhost, :localport # :nodoc:
+  attr_writer :peerhost, :peerhostname, :peerport, :localhost, :localport # :nodoc:
   attr_writer :context # :nodoc:
   attr_writer :ipv # :nodoc:
 
+  def self.rex_gethostbyname(name, resolver = @@resolver)
+    raise ::SocketError.new(
+      "Rex::Socket internal DNS resolution requires passing/setting a resolver"
+    ) unless resolver
+    # Pull both record types
+    v4 = resolver.send(name, ::Net::DNS::A).answer.first
+    v6 = resolver.send(name, ::Net::DNS::AAAA).answer.first
```

Comment:
Forgive me but this seems a little suspicious. What happens if the resolve doesn't respond at all? If it returns an empty  array in that case then this is fine but otherwise if it returns `nil` then this will break.

This also applies to the same pattern in your other function below.

Reply:
I think this pattern comes from the old net/dns code on which i started hacking the DNS horror show way back.
I'm all ears on proposed improvements

Reply:
Ah the joys of old code. I'd probably break this and similar lines down so instead of doing everything on one line we do something like:
1. send
2. Check the response isn't NIL
3. Check the response has an answer with at least one element.
4. Call .first to get the first element.

Reply:
Ah, ok, so this is actually "safe code" - the `Rex::Proto::DNS::CachedResolver`  can _only_  produce a `Dnsruby::Message` when `:send`ing any `String` to the resolver (`pry(Rex::Socket)> v4 = resolver.send('', ::Net::DNS::A).class => Dnsruby::Message`) - so we actually need to verify _name_ not the rest of it for "type safety:"
```ruby
[7] pry(Rex::Socket)> name = 'somerandomthingthatwontresolve.nil'
=> "somerandomthingthatwontresolve.nil"
[8] pry(Rex::Socket)> v4 = resolver.send(name, ::Net::DNS::A).answer.first
=> nil
[9] pry(Rex::Socket)> v4 = resolver.send(name, ::Net::DNS::A).answer
=> []
[10] pry(Rex::Socket)> v4 = resolver.send(name, ::Net::DNS::A).class
=> Dnsruby::Message
```
so thinking
```ruby
  #
  # @param name [String] The hostname to lookup via the resolver
  # @param resolver [Rex::Proto::DNS::CachedResolver] Resolver to query for the name
  # @return [Array] Array mimicking the native getaddrinfo return type
  def self.rex_getaddrinfo(name, resolver = @@resolver)
    raise ::SocketError.new(
      "Rex::Socket internal DNS resolution requires passing/setting a resolver"
    ) unless resolver
    raise ::SocketError.new(
      "Rex::Socket internal DNS resolution requires passing/setting String name to resolve"
    ) unless name.is_a?(String)
```
should do it

Reply:
> Ah, ok, so this is actually "safe code" - the `Rex::Proto::DNS::CachedResolver` can _only_ produce a `Dnsruby::Message` when `:send`ing any `String` to the resolver (`pry(Rex::Socket)> v4 = resolver.send('', ::Net::DNS::A).class => Dnsruby::Message`) - so we actually need to verify _name_ not the rest of it for "type safety:"
> 
> ```ruby
> [7] pry(Rex::Socket)> name = 'somerandomthingthatwontresolve.nil'
> => "somerandomthingthatwontresolve.nil"
> [8] pry(Rex::Socket)> v4 = resolver.send(name, ::Net::DNS::A).answer.first
> => nil
> [9] pry(Rex::Socket)> v4 = resolver.send(name, ::Net::DNS::A).answer
> => []
> [10] pry(Rex::Socket)> v4 = resolver.send(name, ::Net::DNS::A).class
> => Dnsruby::Message
> ```
> 
> so thinking
> 
> ```ruby
>   #
>   # @param name [String] The hostname to lookup via the resolver
>   # @param resolver [Rex::Proto::DNS::CachedResolver] Resolver to query for the name
>   # @return [Array] Array mimicking the native getaddrinfo return type
>   def self.rex_getaddrinfo(name, resolver = @@resolver)
>     raise ::SocketError.new(
>       "Rex::Socket internal DNS resolution requires passing/setting a resolver"
>     ) unless resolver
>     raise ::SocketError.new(
>       "Rex::Socket internal DNS resolution requires passing/setting String name to resolve"
>     ) unless name.is_a?(String)
> ```
> 
> should do it

If I'm not mistaken the code at https://github.com/rapid7/metasploit-framework/pull/16390/files#diff-0e9da515e448f98bae7ee7685586d474b02635232727f8265a0ef1b931281d04R140 shows this could also raise an exception or return `Nil`? So this would actually fail if it returned `nil`.

Reply:
That's fine - https://github.com/rapid7/rex-socket/pull/43/files#diff-4413927154bdecf1370ea1e11874d39c9760d8b8174014ec8dca4dd0798b4e96R901 handles that situation and appropriately raises.

Reply:
I thought `resolver.send(name, ::Net::DNS::A)` would return and then we would call `.answer.select` on a `nil` object which would cause errors?

Reply:
Resolved in https://github.com/rapid7/rex-socket/pull/43/files/21597c92276ea560b05c7cbbfea20df05788ed06..52881ad9232840fe15111b24ac60e9e2e1d2e1d8